lib/foreman_maintain/upgrade_runner.rb

```diff
@@ -0,0 +1,189 @@
+        versions_to_tags.inject([]) do |available_targets, (version, tag)|
+          if !find_scenarios(:tags => [tag]).empty?
+            available_targets << version
+          else
+            available_targets
+          end
+        end
+      end
+
+      def versions_to_tags
+        @versions_to_tags ||= {}
+      end
+
+      # registers target version to specific tag
+      def register_version(version, tag)
+        if versions_to_tags.key?(version) && versions_to_tags[version] != tag
+          raise "Version #{version} already registered to tag #{versions_to_tags[version]}"
+        end
+        @versions_to_tags[version] = tag
+      end
+
+      def clear_register
+        versions_to_tags.lear
+      end
+    end
+
+    attr_reader :version, :tag, :phase
+
+    def initialize(version, reporter, options = {})
+      super(reporter, [], options)
+      @tag = self.class.versions_to_tags[version]
+      raise "Unknown version #{version}" unless tag
+      @version = version
+      @scenario_cache = {}
+      self.phase = :pre_upgrade_checks
+    end
+
+    def scenario(phase)
+      return @scenario_cache[phase] if @scenario_cache.key?(phase)
+      condition = { :tags => [tag, phase] }
+      matching_scenarios = find_scenarios(condition)
+      raise "Too many scenarios match #{condition.inspect}" if matching_scenarios.size > 1
+      @scenario_cache[phase] = matching_scenarios.first
+    end
+
+    def run
+      PHASES.each do |phase|
+        return run_rollback if quit?
+        if skip?(phase)
+          skip_phase(phase)
+        else
+          run_phase(phase)
+        end
+      end
+    end
+
+    def run_rollback
+      # we only are able to rollback from pre_migrations phase
+      if phase == :pre_migrations
+        rollback_pre_migrations
+      end
+    end
+
+    def storage
+      ForemanMaintain.storage("upgrade_#{version}")
+    end
+
+    # serializes the state of the run to storage
+    def save
+      storage[:serialized] = to_hash
+      storage.save
+    end
+
+    # deserializes the state of the run from the storage
+    def load
+      return unless storage[:serialized]
+      load_from_hash(storage[:serialized])
+    end
+
+    def run_phase(phase)
+      with_non_empty_scenario(phase) do |scenario|
+        confirm_scenario(scenario)
+        return if quit?
+        self.phase = phase
+        run_scenario(scenario, false)
+        # if we started from the :pre_upgrade_checks, ensure to ask before
+        # continuing with the rest of the upgrade
+        @ask_to_confirm_upgrade = phase == :pre_upgrade_checks
+      end
+    end
+
+    def skip_phase(skipped_phase)
+      with_non_empty_scenario(skipped_phase) do |scenario|
+        @reporter.before_scenario_starts(scenario)
+        @reporter.puts <<-MESSAGE.strip_heredoc
+          Skipping #{skipped_phase} phase as the last active phase was #{skipped_phase}.
+          To enforce to run the phase, use `upgrade advanced run --phase #{phase}`
```

Comment:
Changing to `Skipping #{skipped_phase} phase as it was already run before.` It can happen, when the migration fails the first time, and one runs the upgrade later to continue.